File: core/src/net/bossdragon/component/render/anim/KeyFrameAnimations.java
package net.bossdragon.component.render.anim;

import com.artemis.PooledComponent;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import net.bossdragon.system.view.render.anim.KeyFrameAnimationsMap;
import net.bossdragon.util.data.Tuple;

/**
 *
 */
public class KeyFrameAnimations extends PooledComponent {
    public Animation<?> currentAnimation = null;

    public KeyFrameAnimationsMap animations = null;
    public float width = 0;
    public float height = 0;
    public float stateTime = 0;

    @Override
    protected void reset() {
        currentAnimation = null;
        animations = null;
        width = height = 0;
        stateTime = 0;
    }

    public KeyFrameAnimations setup(Tuple<String, Animation<?>> animation, Tuple<String, Animation<?>>... otherAnimations) {
        this.animations = new KeyFrameAnimationsMap();
        this.currentAnimation = animation.Item2;

        animations.put(animation.Item1, animation.Item2);
        for (Tuple<String, Animation<?>> tuple : otherAnimations) {
            animations.put(tuple.Item1, tuple.Item2);
        }
        return this;
    }

    public KeyFrameAnimations setup(KeyFrameAnimationsMap animations) {
        this.animations = animations;
        TextureRegion firstFrame = (TextureRegion) animations.get(0).getKeyFrames()[0];
        this.width = firstFrame.getRegionWidth();
        this.height = firstFrame.getRegionHeight();
        currentAnimation = this.animations.get(0);

        return this;
    }

    public boolean setAnimation(int index) {
        final Animation<?> newAnim = animations.get(index);
        final boolean isChanging = newAnim != currentAnimation;

        if (isChanging)
            currentAnimation = animations.get(index);

        return isChanging;
    }

    public boolean setAnimation(String name) {
        final Animation<?> newAnim = animations.get(name);
        final boolean isChanging = newAnim != currentAnimation;

        if (isChanging)
            currentAnimation = animations.get(name);

        return isChanging;
    }
}

File: core/src/net/bossdragon/component/render/anim/KeyFrameAnimations.kt
package net.bossdragon.component.render.anim

import com.artemis.PooledComponent
import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.TextureRegion
import net.bossdragon.system.view.render.anim.KeyFrameAnimationsMap
import net.bossdragon.util.data.Tuple

/**
 * This component needs {@link Progressing} component!
 */
class KeyFrameAnimations : PooledComponent() {
    var currentAnimation: Animation<*>? = null
        private set

    @JvmField var animations: KeyFrameAnimationsMap? = null
    var width: Float = 0.toFloat()
    var height: Float = 0.toFloat()
    var stateTime: Float = 0f


    init {
        this.animations = KeyFrameAnimationsMap()
    }

    fun setup(animation: Tuple<String, Animation<*>>, vararg otherAnimations: Tuple<String, Animation<*>>): KeyFrameAnimations {
        this.animations = KeyFrameAnimationsMap()
        this.currentAnimation = animation.Item2

        animations!!.put(animation.Item1, animation.Item2)
        for (tuple in otherAnimations) {
            animations!!.put(tuple.Item1, tuple.Item2)
        }
        return this
    }

    fun setup(animations: KeyFrameAnimationsMap): KeyFrameAnimations {
        this.animations = animations
        val firstFrame = animations.get(0).keyFrames[0] as TextureRegion
        this.width = firstFrame.getRegionWidth().toFloat()
        this.height = firstFrame.getRegionHeight().toFloat()
        currentAnimation = this.animations!![0]

        return this
    }

    fun setAnimation(index: Int): Boolean {
        val newAnim = animations!![index]
        val isChanging = newAnim != currentAnimation

        currentAnimation = newAnim
        return isChanging
    }

    fun setAnimation(name: String): Boolean {
        val newAnim = animations!![name]
        val isChanging = newAnim != currentAnimation

        currentAnimation = animations!![name]
        return isChanging
    }


    override fun reset() {
        currentAnimation = null
        animations = null
        height = 0f
        width = 0f
    }
}
